### PR TITLE
log: allow different log levels

### PR DESCRIPTION
**Pull request by jmrodri**
With this feature we can use --zap-level 10 and this will allow
log.V(10).Info("hello world") to be seen in the logs. Prior to this
change debug (log.V(1)) was the lowest value attainable.

We also disable sampling when using an integer less than -1
(i.e. beyond debug). The zap sampling code assumes the lowest number for
sampling is DebugLevel (-1) which isn't the case in this scenario.
Disabling sampling makes sense at this point because sampling is used
for better performing logs which isn't required for trace level type
logs.

We also handle negative values being passed in.

### CODE REVIEW

**Comment by jmrodri** [comment]
I updated proxy.go to have 6 levels of log to test this PR out. Here is the output:
```
[jesusr@transam foo-operator{master}]$ operator-sdk up local --zap-level 10
INFO[0000] Running the operator locally.                
INFO[0000] Using namespace myproject.                   
{"level":"info","ts":1552588792.6807225,"logger":"cmd","msg":"Go Version: go1.11.5"}
{"level":"info","ts":1552588792.6807659,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1552588792.6807742,"logger":"cmd","msg":"Version of operator-sdk: v0.5.0+git"}
{"level":"info","ts":1552588792.680802,"logger":"cmd","msg":"Watching namespace.","Namespace":"myproject"}
{"level":"info","ts":1552588792.7752655,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1552588792.7754438,"logger":"leader","msg":"Skipping leader election; not running in a cluster."}
{"level":"info","ts":1552588792.7771642,"logger":"proxy","msg":"Starting to serve","Address":"127.0.0.1:8888"}
{"level":"debug","ts":1552588792.7773304,"logger":"proxy","msg":"Starting to serve level 1","Address":"127.0.0.1:8888"}
{"level":"Level(-2)","ts":1552588792.7773762,"logger":"proxy","msg":"Starting to serve level 2","Address":"127.0.0.1:8888"}
{"level":"Level(-3)","ts":1552588792.7774038,"logger":"proxy","msg":"Starting to serve level 3","Address":"127.0.0.1:8888"}
{"level":"Level(-4)","ts":1552588792.7774181,"logger":"proxy","msg":"Starting to serve level 4","Address":"127.0.0.1:8888"}
{"level":"Level(-5)","ts":1552588792.7774332,"logger":"proxy","msg":"Starting to serve level 5","Address":"127.0.0.1:8888"}
{"level":"Level(-6)","ts":1552588792.7774472,"logger":"proxy","msg":"Starting to serve level 6","Address":"127.0.0.1:8888"}
{"level":"error","ts":1552588792.7774608,"logger":"proxy","msg":"Starting to serve error level","Address":"127.0.0.1:8888","stacktrace":"github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\tsrc/github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/operator-framework/operator-sdk/pkg/ansible/proxy.Run.func2\n\tsrc/github.com/operator-framework/operator-sdk/pkg/ansible/proxy/proxy.go:403"}
{"level":"info","ts":1552588792.7777417,"logger":"manager","msg":"Using default value for workers 1"}
{"level":"info","ts":1552588792.777771,"logger":"ansible-controller","msg":"Watching resource","Options.Group":"cache.example.com","Options.Version":"v1alpha1","Options.Kind":"FooService"}
{"level":"info","ts":1552588792.7781951,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"fooservice-controller","source":"kind source: cache.example.com/v1alpha1, Kind=FooService"}
{"level":"info","ts":1552588792.8786395,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"fooservice-controller"}
{"level":"info","ts":1552588792.9789739,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"fooservice-controller","worker count":1}
```

**Comment by jmrodri** [comment]
Using log level 3 notice that 4, 5, & 6 are no longer visible.
```
[jesusr@transam foo-operator{master}]$ operator-sdk up local --zap-level 3
INFO[0000] Running the operator locally.                
INFO[0000] Using namespace myproject.                   
{"level":"info","ts":1552588862.6171176,"logger":"cmd","msg":"Go Version: go1.11.5"}
{"level":"info","ts":1552588862.6171565,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1552588862.6171675,"logger":"cmd","msg":"Version of operator-sdk: v0.5.0+git"}
{"level":"info","ts":1552588862.6171973,"logger":"cmd","msg":"Watching namespace.","Namespace":"myproject"}
{"level":"info","ts":1552588862.710273,"logger":"leader","msg":"Trying to become the leader."}
{"level":"info","ts":1552588862.7104235,"logger":"leader","msg":"Skipping leader election; not running in a cluster."}
{"level":"info","ts":1552588862.712774,"logger":"proxy","msg":"Starting to serve","Address":"127.0.0.1:8888"}
{"level":"debug","ts":1552588862.7128866,"logger":"proxy","msg":"Starting to serve level 1","Address":"127.0.0.1:8888"}
{"level":"Level(-2)","ts":1552588862.7129002,"logger":"proxy","msg":"Starting to serve level 2","Address":"127.0.0.1:8888"}
{"level":"Level(-3)","ts":1552588862.7129188,"logger":"proxy","msg":"Starting to serve level 3","Address":"127.0.0.1:8888"}
{"level":"error","ts":1552588862.7129462,"logger":"proxy","msg":"Starting to serve error level","Address":"127.0.0.1:8888","stacktrace":"github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\tsrc/github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/operator-framework/operator-sdk/pkg/ansible/proxy.Run.func2\n\tsrc/github.com/operator-framework/operator-sdk/pkg/ansible/proxy/proxy.go:403"}
{"level":"info","ts":1552588862.7132838,"logger":"manager","msg":"Using default value for workers 1"}
{"level":"info","ts":1552588862.7133071,"logger":"ansible-controller","msg":"Watching resource","Options.Group":"cache.example.com","Options.Version":"v1alpha1","Options.Kind":"FooService"}
{"level":"info","ts":1552588862.7137141,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"fooservice-controller","source":"kind source: cache.example.com/v1alpha1, Kind=FooService"}
{"level":"info","ts":1552588862.8141239,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"fooservice-controller"}
{"level":"info","ts":1552588862.9144537,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"fooservice-controller","worker count":1}
```

File: pkg/log/zap/logger.go [hunk 1]
```diff
@@ -75,6 +75,13 @@ func getConfig(destWriter io.Writer) config {
 	}
 	if levelVal.set {
 		c.level = zap.NewAtomicLevelAt(levelVal.level)
+
+		// Disable sampling when we are in debug mode. Otherwise, this will
+		// cause index out of bounds errors in the sampling code.
```
Comment (joelanford) [inline pkg/log/zap/logger.go]: Is this the fix for the bug mentioned [here](https://github.com/operator-framework/operator-sdk/pull/873#discussion_r252335389)?

Comment (shawn-hurley) [reply pkg/log/zap/logger.go]: Yeah, I bet that this the case.

Comment (joelanford) [reply pkg/log/zap/logger.go]: I'm pretty sure it is. Nice find @jmrodri!

With this being a possible source of runtime panics, can we add some unit tests to make sure we don't introduce a regression later?

File: pkg/log/zap/flags.go [hunk 2]
```diff
@@ -91,11 +91,28 @@ type levelValue struct {
...
+		i, err := strconv.Atoi(lower)
+		if err != nil {
+			return fmt.Errorf("invalid log level \"%s\"", l)
+		}
+
+		if i > 0 {
+			lvl = -1 * i
+		} else {
+			lvl = i
+		}
```
Comment (joelanford) [inline pkg/log/zap/flags.go]: In the flag description, we say zap-level can be any integer value greater than 0, but here we allow all integers. So I think we probably need to change the flag description or this logic.

**TL;DR: I think we should allow only logr-supported levels (`error`, `info`, `debug`, and non-negative integers), and anything else should result in an error.**

Mixing zap and logr confuses things because zap uses negative numbers for increasing levels of verbosity, whereas [logr specifically forbids negative numbers](https://github.com/go-logr/logr/blob/9fb12b3b21c5415d16ac18dc5cd42c1cfdd40c4e/logr.go#L136-L139). Zap also has levels that aren't supported by logr.

Operator _developers_ are dealing with the logr interface, and are probably more likely to think to use positive numbers with this flag to denote higher verbosity.

Operator _users_ are seeing only the flag name (`--zap-level`), and are probably more likely to think to use negative numbers since that's what zap uses natively.

I feel like allowing both positive and negative numbers could be confusing to both developers and users. There was a [decent amount of discussion](https://github.com/operator-framework/operator-sdk/pull/873#discussion_r251292941) in the original PR about what the acceptable integer range and string values would be.

My initial thought was that we should use zap's convention for the flag values since that's the exposed flag name:
- `error`
- `info`/`0`
- `debug`/`-1`
- `-2`
- `-3`
- ...

The reason I ended up recommending the logr convention in the original PR was that using the zap convention could give the impression that all zap levels (`warn`, `panic`, `fatal`, etc) are supported. And since zapr doesn't fail when you pass it the negative numbers associated with those levels through the logr interface, we'll probably get users trying to do something not supported by logr. 

By sticking with only logr-supported levels, it should prevent situations where an operator developer uses zapr's unsupported negative log levels in his code, and later changes to a different logr implementation that panics with those negative numbers.

Comment (jmrodri) [reply pkg/log/zap/flags.go]: Updated test and made negative values return error.

File: pkg/log/zap/logger.go [hunk 3]
```diff
@@ -75,6 +75,13 @@ func getConfig(destWriter io.Writer) config {
...
+		if levelVal.level < -1 {
+			sampleVal.set = false
+			c.sample = false
```
Comment (joelanford) [inline pkg/log/zap/logger.go]: Do we think it's okay to silently disable sampling when `--zap-sample=true`?

Instead, we could `log.Fatal()` or change the API to allow us to return an error. Thoughts?

File: pkg/log/zap/logger_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		t.Run(tc.name, func(t *testing.T) {
+			development = tc.inDevel
+			encoderVal = tc.inEncoder
+			levelVal = tc.inLevel
+			sampleVal = tc.inSample
+
+			cfg := getConfig()
+			assert.Equal(t, tc.expected.level, cfg.level)
+			assert.Equal(t, len(tc.expected.opts), len(cfg.opts))
+			assert.Equal(t, tc.expected.sample, cfg.sample)
```
Comment (joelanford) [inline pkg/log/zap/logger_test.go]: Should we also actually build the logger from the config and attempt to call `log.V(10).Info(...)` to make sure we don't panic when debug logging is enabled?

File: pkg/log/zap/logger.go (hunk 3 above)
Comment (shawn-hurley) [reply pkg/log/zap/logger.go]: Yes, I do. 

We should document that this is what we do, but TBH how many users are purposefully turning on sampling (which is for log performance in production) AND setting the log level to trace level logs. I would assume this is a mistake and we should probably just make the right decision.

File: pkg/log/zap/logger_test.go (hunk 4 above)
Comment (shawn-hurley) [reply pkg/log/zap/logger_test.go]: That seems like an e2e test. the unit test is testing the code is behaving how we expected. 

I could see us enabling this in the ansible e2e test to help debug CI failures. I would imagine this would be a follow-up PR. 

thoughts?

File: pkg/log/zap/logger.go (hunk 3 above)
Comment (joelanford) [reply pkg/log/zap/logger.go]: Also, this is more of a nit, but I'm a little nervous about this check happening in the `if levelVal.set { ... }` block. I know we have unit tests now, but if we ever change our default log level to debug, this check would get skipped.

It would probably be better to use `levelVal.level` to set the default in line 69 and then do another check in the `if sampleVal.set { ... }` block to make sure we turn sampling back on when debug logging is enabled.

File: pkg/log/zap/logger_test.go (hunk 4 above)
Comment (joelanford) [reply pkg/log/zap/logger_test.go]: It seems like something we could do pretty easily in a `func TestLoggerDoesntPanic(t *testing.T) {}`-style test, and I think such a test would fall under the umbrella of "code is behaving how we expected".

But I'm on board with doing something like this as a follow up.

File: pkg/log/zap/logger.go (hunk 3 above)
Comment (shawn-hurley) [reply pkg/log/zap/logger.go]: @jmrodri I think we should use c.level to determine this on line 89.

Comment (jmrodri) [reply pkg/log/zap/logger.go]: I moved the if block to out of the `levelVal.set` to just before return c.

```golang
func getConfig() config {                                                            
    var c config                                                                     
                                                                                     
    // Set the defaults depending on the log mode (development vs. production)       
    if development {                                                                 
        c.encoder = consoleEncoder()                                                 
        c.level = zap.NewAtomicLevelAt(zap.DebugLevel)                               
        c.opts = append(c.opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel))
        c.sample = false                                                             
    } else {                                                                         
        c.encoder = jsonEncoder()                                                    
        c.level = zap.NewAtomicLevelAt(zap.InfoLevel)                                
        c.opts = append(c.opts, zap.AddStacktrace(zap.WarnLevel))                    
        c.sample = true                                                              
    }                                                                                
                                                                                     
    // Override the defaults if the flags were set explicitly on the command line    
    if encoderVal.set {                                                              
        c.encoder = encoderVal.encoder                                               
    }                                                                                
    if levelVal.set {                                                                
        c.level = zap.NewAtomicLevelAt(levelVal.level)                               
                                                                                     
    }                                                                                
    if sampleVal.set {                                                               
        c.sample = sampleVal.sample                                                  
    }                                                                                
                                                                                     
    // Disable sampling when we are in debug mode. Otherwise, this will              
    // cause index out of bounds errors in the sampling code.                        
    if c.level.Level() < -1 {                                                        
        sampleVal.set = false                                                        
        c.sample = false                                                             
    }                                                                                
    return c                                                                         
}                                                                                    
```

Comment (jmrodri) [reply pkg/log/zap/logger.go]: I moved the if block to just before we return the config.

```golang
func getConfig() config {                                                             
    var c config                                                                      
                                                                                      
    // Set the defaults depending on the log mode (development vs. production)        
    if development {                                                                  
        c.encoder = consoleEncoder()                                                  
        c.level = zap.NewAtomicLevelAt(zap.DebugLevel)                                
        c.opts = append(c.opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel)) 
        c.sample = false                                                              
    } else {                                                                          
        c.encoder = jsonEncoder()                                                     
        c.level = zap.NewAtomicLevelAt(zap.InfoLevel)                                 
        c.opts = append(c.opts, zap.AddStacktrace(zap.WarnLevel))                     
        c.sample = true                                                               
    }                                                                                 
                                                                                      
    // Override the defaults if the flags were set explicitly on the command line     
    if encoderVal.set {                                                               
        c.encoder = encoderVal.encoder                                                
    }                                                                                 
    if levelVal.set {                                                                 
        c.level = zap.NewAtomicLevelAt(levelVal.level)                                
                                                                                      
    }                                                                                 
    if sampleVal.set {                                                                
        c.sample = sampleVal.sample                                                   
    }                                                                                 
                                                                                      
    // Disable sampling when we are in debug mode. Otherwise, this will               
    // cause index out of bounds errors in the sampling code.                         
    if c.level.Level() < -1 {                                                         
        c.sample = false                                                              
    }                                                                                 
    return c                                                                          
}                                                                                     
```

File: pkg/log/zap/logger_test.go (hunk 4 above)
Comment (jmrodri) [reply pkg/log/zap/logger_test.go]: @joelanford @shawn-hurley I can add a more thorough test in a follow up if that is acceptable.

File: pkg/log/zap/logger_test.go [hunk 5]
```diff
@@ -166,6 +171,22 @@ func TestGetConfig(t *testing.T) {
...
 			assert.Equal(t, len(tc.expected.opts), len(cfg.opts))
 			assert.Equal(t, tc.expected.sample, cfg.sample)
+
+			dalog := createLogger(cfg, os.Stderr)
+			dalog.V(10).Info("this should not panic")
 		})
 	}
 }
+
+func createLogger(cfg config, dest io.Writer) logr.Logger {
```
Comment (joelanford) [inline pkg/log/zap/logger_test.go]: We could probably put this in `logger.go` and use it from our test and from `LoggerTo` so we don't have to duplicate it.

Comment (jmrodri) [reply pkg/log/zap/logger_test.go]: I copied it from logger.go. I'm not 100% sure what all the options are to the logger, I'd probably like to remove a bunch of those. I also don't like putting code that I need in a test in the mainline code base unless it is going to be used generally.  So -1 to putting `createLogger` in `logger.go`

**Comment by jmrodri** [comment]
I now create a logger in the test to ensure it doesn't cause a panic. With the unaltered code:
```golang
    ...
     // Disable sampling when we are in debug mode. Otherwise, this will 
     // cause index out of bounds errors in the sampling code.           
     if c.level.Level() < -1 {                                           
         c.sample = false                                                
     }                                                                   
     return c                                                            
 }                                                                       
```

When I run the test, the log works just fine.
```
[jesusr@transam zap{log-levels}]$ go test
{"level":"Level(-10)","ts":1552667568.3941715,"msg":"this should not panic"}                                        
{"level":"Level(-10)","ts":1552667568.3942807,"msg":"this should not panic"}                                        
PASS
ok      github.com/operator-framework/operator-sdk/pkg/log/zap  0.009s  
```

Now the part I really liked which is why @joelanford suggested it, if I comment out the code that disables sampling, we detect the panic.

```golang
...
    // Disable sampling when we are in debug mode. Otherwise, this will 
    // cause index out of bounds errors in the sampling code.           
    // if c.level.Level() < -1 {                                        
    //     c.sample = false                                             
    // }                                                                
    return c                                                            
}      
```
Using the above, we can see the test will panic indicating that clearly we broke something :)
```
        [jesusr@transam zap{log-levels}]$ go test
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range

goroutine 42 [running]:
testing.tRunner.func1(0xc000115900)
        /usr/lib/golang/src/testing/testing.go:792 +0x387
panic(0x8ba720, 0xd49470)
        /usr/lib/golang/src/runtime/panic.go:513 +0x1b9
github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/zapcore.(*counters).get(0xc000306000, 0x95bcf6, 0x9
3d902, 0x15, 0xc000115400)
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/zapcore/sampler.
go:48 +0x79
github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/zapcore.(*sampler).Check(0xc000181e60, 0xf6, 0xbf1b
12d51e5e730a, 0x62477e, 0xd57520, 0x0, 0x0, 0x93d902, 0x15, 0x0, ...)                                               
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/zapcore/sampler.
go:128 +0x9d
github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap.(*Logger).check(0xc0000d31a0, 0xf6, 0x93d902, 0x15,
 0xc0000cd89c)
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/logger.go:269 +0
x137
github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap.(*Logger).Check(0xc0000d31a0, 0xf6, 0x93d902, 0x15,
 0xc00019e420)
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/vendor/go.uber.org/zap/logger.go:172 +0
x48
github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr.(*infoLogger).Info(0xc00019e420, 0x93d902,
0x15, 0x0, 0x0, 0x0)
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/vendor/github.com/go-logr/zapr/zapr.go:
69 +0x4e
github.com/operator-framework/operator-sdk/pkg/log/zap.TestGetConfig.func1(0xc000115900)                            
        /home/devel/jesusr/dev/src/github.com/operator-framework/operator-sdk/pkg/log/zap/logger_test.go:183 +0x4d5 
testing.tRunner(0xc000115900, 0xc00019e360)
        /usr/lib/golang/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
        /usr/lib/golang/src/testing/testing.go:878 +0x35c
exit status 2
FAIL    github.com/operator-framework/operator-sdk/pkg/log/zap  0.013s   
```                                                         

Thanks for the suggestion @joelanford

File: pkg/log/zap/logger_test.go (hunk 5 above)
Comment (joelanford) [reply pkg/log/zap/logger_test.go]: Sorry, what I meant was that most of these same lines exist in `LoggerTo`

https://github.com/operator-framework/operator-sdk/blob/720a33546ec1c5d05a1c93e73656fa624b1617ca/pkg/log/zap/logger.go#L33-L47

For `LoggerTo`, we should do something like this:
```go
func LoggerTo(destWriter io.Writer) logr.Logger {
	conf := getConfig()
	return createLogger(config, destWriter)
}

func createLogger(cfg conf, destWriter io.Writer) logr.Logger {
	syncer := zapcore.AddSync(destWriter)
	conf.encoder = &logf.KubeAwareEncoder{Encoder: conf.encoder, Verbose: conf.level.Level() < 0}
	if conf.sample {
		conf.opts = append(conf.opts, zap.WrapCore(func(core zapcore.Core) zapcore.Core {
			return zapcore.NewSampler(core, time.Second, 100, 100)
		}))
	}
	conf.opts = append(conf.opts, zap.AddCallerSkip(1), zap.ErrorOutput(syncer))
	log := zap.New(zapcore.NewCore(conf.encoder, syncer, conf.level))
	log = log.WithOptions(conf.opts...)
	return zapr.NewLogger(log)
}
```

File: pkg/log/zap/logger.go [hunk 6]
```diff
@@ -75,10 +75,16 @@ func getConfig(destWriter io.Writer) config {
 	}
 	if levelVal.set {
 		c.level = zap.NewAtomicLevelAt(levelVal.level)
+
```
Comment (joelanford) [inline pkg/log/zap/logger.go]: Nit: this line is extraneous

File: pkg/log/zap/flags.go [hunk 7]
```diff
@@ -37,8 +37,8 @@ func init() {
 	zapFlagSet = pflag.NewFlagSet("zap", pflag.ExitOnError)
 	zapFlagSet.BoolVar(&development, "zap-devel", false, "Enable zap development mode (changes defaults to console encoder, debug log level, and disables sampling)")
 	zapFlagSet.Var(&encoderVal, "zap-encoder", "Zap log encoding ('json' or 'console')")
-	zapFlagSet.Var(&levelVal, "zap-level", "Zap log level (one of 'debug', 'info', 'error')")
-	zapFlagSet.Var(&sampleVal, "zap-sample", "Enable zap log sampling")
+	zapFlagSet.Var(&levelVal, "zap-level", "Zap log level (one of 'debug', 'info', 'error' or any integer value > 0)")
+	zapFlagSet.Var(&sampleVal, "zap-sample", "Enable zap log sampling. Sampling will be disabled for log levels > 0")
```
Comment (joelanford) [inline pkg/log/zap/flags.go]: To match our [check in `getConfig`](https://github.com/operator-framework/operator-sdk/pull/1210/files#diff-4407315726717f30063d558928ad9188R86), I'd suggest this change to the wording:
```suggestion
	zapFlagSet.Var(&sampleVal, "zap-sample", "Enable zap log sampling. Sampling will be disabled for integer log levels > 1")
```

File: pkg/log/zap/logger.go (hunk 6 above)
Comment (jmrodri) [reply pkg/log/zap/logger.go]: That's not new code though. That was there from before :)

File: pkg/log/zap/flags.go (hunk 7 above)
Comment (jmrodri) [reply pkg/log/zap/flags.go]: Changed it.
```
[jesusr@transam foo-operator{master}]$ operator-sdk up local --help
The operator-sdk up local command launches the operator on the local machine
by building the operator binary with the ability to access a
kubernetes cluster using a kubeconfig file.

Usage:
  operator-sdk up local [flags]

Flags:
      --go-ldflags string           Set Go linker options
  -h, --help                        help for local
      --inject-owner-ref            (ansible operator) The ansible operator will inject owner references unless this flag is false (default true)
      --kubeconfig string           The file path to kubernetes configuration file; defaults to location specified by $KUBECONFIG with a fallback to $HOME/.kube/config if not set
      --max-workers int             (ansible operator) Maximum number of workers to use. Overridden by environment variable. (default 1)
      --namespace string            The namespace where the operator watches for changes.
      --operator-flags string       The flags that the operator needs. Example: "--flag1 value1 --flag2=value2"
      --reconcile-period duration   (ansible operator) Default reconcile period for controllers (default 1m0s)
      --watches-file string         (ansible operator) Path to the watches file to use (default "./watches.yaml")
      --zap-devel                   Enable zap development mode (changes defaults to console encoder, debug log level, and disables sampling)
      --zap-encoder encoder         Zap log encoding ('json' or 'console')
      --zap-level level             Zap log level (one of 'debug', 'info', 'error' or any integer value > 1) (default info)
      --zap-sample sample           Enable zap log sampling. Sampling will be disabled for integer log levels > 1
```

File: pkg/log/zap/logger.go (hunk 6 above)
Comment (joelanford) [reply pkg/log/zap/logger.go]: Are you sure? This is what it looks like on master (no blank line between setting `c.level` and the closing bracket): https://github.com/operator-framework/operator-sdk/blob/c6e209a42be7d4ef6a5b1077cfafea7fc6f7a77a/pkg/log/zap/logger.go#L76-L78

I don't think we'd see the + green line otherwise. I'm guessing it was just left over from moving the disable sampling logic.

File: pkg/log/zap/logger_test.go (hunk 5 above)
Comment (jmrodri) [reply pkg/log/zap/logger_test.go]: I disagree. Because in the test I don't actually set the encoder to a KubeAwareEncoder. Also, while I did copy the other stuff verbatim. I can see us trimming down the call used in the test to something smaller.

Comment (jmrodri) [reply pkg/log/zap/logger_test.go]: So I still say -1 to adding a `createLogger` to logger.go. I think they just LOOK the same but I still believe that we will need to be more flexible in how we create the loggers in testing.

File: pkg/log/zap/logger.go (hunk 6 above)
Comment (jmrodri) [reply pkg/log/zap/logger.go]: Upon further review, that line is NOT extraneous. Without it the level will never be set correctly. It will either be debug or warn. We'd have to change the else block of the development check. But if we're doing that we'd be checking to see if levelVal was set before setting the level. And if it wasn't set we'd default to WarnLevel. At which point we'd just be moving this line to a different place. So with all due respect, I think it should remain here for this PR.

File: pkg/log/zap/logger_test.go (hunk 5 above)
Comment (joelanford) [reply pkg/log/zap/logger_test.go]: IMO we should be testing what our public API is using, so I don't think our test should use an alternate method to create the logger. I could set `conf.sample` in `LoggerTo` and break things without tripping the unit test.
